### PR TITLE
36 Improve output and verbose options

### LINKED ISSUE #36: Improve output and verbose levels
Currently, some output is always given. In order to reduce the output to a minimum / zero, we should have a verbose level which suppresses all output. Furthermore, the verbose levels should be adapted consistently.

### PR DESCRIPTION
**Pull request by mknaranja**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] New code adheres with the [coding guidelines](https://github.com/mknaranja/GMGPolar/wiki)
* [x] No large data files have been added to the repository. Maximum size for files should be of the order of KB not MB. In particular avoid adding of pdf, word, or other files that cannot be change-tracked correctly by git.

If functions were changed or functionality was added:
* [ ] Tests for new functionality has been added
* [x] A local test was succesful

If new functionality was added:
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [ ] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [ ] Are new methods referenced? Did you provide further documentation?

[ ] The following questions are addressed in the documentation (if need be): 
* Developers (what did you do?, how can it be maintained?)
* For users (how to use your work?)
* For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [ ] Coverage report for new code is acceptable. 
* [ ] No large data files have been added to the repository. Maximum size for files should be of the order of KB not MB. In particular avoid adding of pdf, word, or other files that cannot be change-tracked correctly by git.

Closes #36

### CODE REVIEW

**Comment by mknaranja** [comment]
@Philou31 Could you have a quick look if you like my suggestion? Better ideas or modifications proposed? How about the many commented lines of output (in particular disp()). Do we remove or uncomment and integrate them?

@CodingAllan Can you have a check that I did not introduce any new error?

File: src/polar_multigrid.cpp [hunk 1]
```diff
@@ -586,23 +591,23 @@ void gmgpolar::prepare_op_levels()
                 v_level[l]->define_nz_P();
 
                 // // Number of nonzeros in the matrix
-                if (gyro::icntl[Param::verbose] > 2)
+                if (gyro::icntl[Param::verbose] > 3)
                     std::cout << "nz_P: " << v_level[l]->nz_P << "\n";
```
Comment (CodingAllan) [inline src/polar_multigrid.cpp:595]: doesnt this fit better into "print out all kinds of variable values" ? so gyro::icntl[Param::verbose]>5

File: src/polar_multigrid.cpp [hunk 2]
```diff
@@ -586,23 +591,23 @@ void gmgpolar::prepare_op_levels()
...
                     std::cout << "nz_P_inj: " << v_level[l]->nz_P_inj << "\n";
                 v_level[l]->ri_prol_inj = std::vector<int>(v_level[l]->nz_P_inj);
                 v_level[l]->ci_prol_inj = std::vector<int>(v_level[l]->nz_P_inj);
                 v_level[l]->v_prol_inj  = std::vector<double>(v_level[l]->nz_P_inj);
                 v_level[l]->build_prolongation_inj();
 
                 // // Number of nonzeros in the matrix
-                if (gyro::icntl[Param::verbose] > 2)
+                if (gyro::icntl[Param::verbose] > 3)
                     std::cout << "nz_P_ex: " << v_level[l]->nz_P_ex << "\n";
```
Comment (CodingAllan) [inline src/polar_multigrid.cpp:611]: doesnt this fit better into "print out all kinds of variable values" ? so gyro::icntl[Param::verbose]>5

File: src/multigrid_iter.cpp [hunk 3]
```diff
@@ -181,7 +181,7 @@ void gmgpolar::multigrid_iter()
             double error_difference = fabs(nrm_2_err[it] - nrm_2_err[it - 1]);
             convergence_criterium   = error_difference / nrm_2_err[0];
 
-            if (gyro::icntl[Param::verbose] > 0)
+            if (gyro::icntl[Param::verbose] > 1)
                 std::cout << "--> Iteration " << it << ": error norm = " << nrm_2_err[it]
                           << ", relative error = " << convergence_criterium << std::endl;
         }
```
Comment (CodingAllan) [inline src/multigrid_iter.cpp:187]: before Line 192 (below) shoudnt we set gyro::icntl[Param::verbose]>0 ? the if-statement is commented out

File: src/multigrid_iter.cpp [hunk 4]
```diff
@@ -224,14 +224,14 @@ void gmgpolar::multigrid_iter()
         std::cout << "2-norm of error = " << nrm_2 << std::endl;
```
Comment (CodingAllan) [inline src/multigrid_iter.cpp:224]: gyro::icntl[Param::verbose]>0.
Or maybe Param::check_error should "overwrite" verbose =0 in this case. Then this is fine I think.

File: src/polar_multigrid.cpp (hunk 1 above)
Comment (mknaranja) [reply src/polar_multigrid.cpp:595]: I don't think so. nz_P is not a variable but an information on the problem setting.

File: src/polar_multigrid.cpp [hunk 5]
```diff
@@ -586,23 +591,23 @@ void gmgpolar::prepare_op_levels()
...
                     std::cout << "nz_P: " << v_level[l]->nz_P << "\n";
                 v_level[l]->ri_prol = std::vector<int>(v_level[l]->nz_P);
                 v_level[l]->ci_prol = std::vector<int>(v_level[l]->nz_P);
                 v_level[l]->v_prol  = std::vector<double>(v_level[l]->nz_P);
                 v_level[l]->build_prolongation_bi();
 
                 // // Number of nonzeros in the matrix
-                if (gyro::icntl[Param::verbose] > 2)
+                if (gyro::icntl[Param::verbose] > 3)
                     std::cout << "nz_P_inj: " << v_level[l]->nz_P_inj << "\n";
```
[the remark this answers is not in the document]
Comment (mknaranja) [reply src/polar_multigrid.cpp:603]: see above

File: src/gyro.cpp [hunk 6]
```diff
@@ -243,7 +243,7 @@ void gyro::show_params()
  *
  * Computes the distance of a node (x, y) to the Dirichlet boundary, using 
  * tolerance of 1e-10 to define boundary.
- * Highly dependent on dcntl[Param::r0_DB] which is 1e6 is icntl[Param::DirBC_Interior]=0.
+ * Highly dependent on dcntl[Param::r0_DB] which should be very small if icntl[Param::DirBC_Interior]=0.
  * 
  * \param x: the x coordinate of the node
  * \param y: the  coordinate of the node
```
Comment (CodingAllan) [inline src/gyro.cpp:249]: in gyro::distBoundary, line 267 maybe one can add that to verbose > 5.

this function is only used in the non-optimized construction of A so also not used that regulary.

File: src/create_grid_polar.cpp [hunk 7]
```diff
@@ -243,7 +243,9 @@ void level::build_theta()
             theta[i] = fac * i;
     }
     else {
-        std::cout << "Anisotropy chosen in theta.\n";
+        if (gyro::icntl[Param::verbose] > 2) {
+            std::cout << "Anisotropy chosen in theta.\n";
+        }
         ntheta     = pow(2, ceil(log2(nr)) - 1);
```
Comment (CodingAllan) [inline src/create_grid_polar.cpp:249]: ntheta only depends on nr and not on ntheta_exp.

Further, when we calculate "delete circles" in level::define_line_splitting() (smoother.cpp line 73), we use that the grid spacing in theta is constant, dont we ? If anisotrophy is chosen in theta I think one should update that function as well!

File: src/gyro.cpp (hunk 6 above)
Comment (mknaranja) [reply src/gyro.cpp:249]: well spotted!

File: src/level.cpp [hunk 8]
```diff
@@ -146,10 +146,11 @@ void level::build_bound()
...
     }
 
-    if (gyro::icntl[Param::verbose] > 3)
+    if (gyro::icntl[Param::verbose] > 5)
+        std::cout << "Printing for (r, theta) if the node is on the boundary.\n";
         for (int j = 0; j < nr; j++)
             for (int i = 0; i < ntheta_int; i++)
-                std::cout << "DISTBOUNDARY (" << r[j] << ", " << theta[j] << "): " << is_bound[j * ntheta_int + i]
+                std::cout << "(" << r[j] << ", " << theta[j] << "): " << is_bound[j * ntheta_int + i]
                           << "\n";
```
[the remark this answers is not in the document]
Comment (mknaranja) [reply src/level.cpp:154]: I am not sure here. It is called in `polar_multigrid.cpp`. But yes, there had been a change (using simplifications through assumptions) in computation of the boundary nodes from Matlab to C++

File: src/create_grid_polar.cpp (hunk 7 above)
Comment (mknaranja) [reply src/create_grid_polar.cpp:249]: https://github.com/mknaranja/GMGPolar/issues/42

File: src/define_coarse_nodes.cpp [hunk 9]
```diff
@@ -51,13 +51,13 @@ void gmgpolar::define_coarse_nodes()
...
     for (int i = 0; i < levels; i++) {
-        if (gyro::icntl[Param::verbose] > 1)
+        if (gyro::icntl[Param::verbose] > 3)
             std::cout << "=> Defining coarse nodes on level " << i << "\n";
         v_level[i]->store_theta_n_co();
         if (i < levels - 1) {
             v_level[i]->define_coarse_nodes_onelevel(v_level[i + 1]);
         }
-        if (gyro::icntl[Param::verbose] > 1) {
+        if (gyro::icntl[Param::verbose] > 3) {
```
[the remark this answers is not in the document]
Comment (mknaranja) [reply src/define_coarse_nodes.cpp:60]: No, I think >3 is okay. Thanks for checking anyhow :)

File: src/level.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mknaranja) [reply src/level.cpp:154]: Now in #43

**Review by Philou31 — APPROVED** [review]

**Pull request merged**